tests/test_chaiverse/test_metrics.py

```diff
@@ -40,9 +40,10 @@ def test_developer_can_call_get_submission_metrics_and_pass_in_developer_key_as_
 
 @mock.patch('chaiverse.metrics.get_all_historical_submissions')
 @mock.patch('chaiverse.utils.guanaco_data_dir')
+@mock.patch('chaiverse.utils._save_to_cache')
 @freeze_time('2023-07-28 00:00:00')
 @vcr.use_cassette(os.path.join(RESOURCE_DIR, 'test_get_raw_leaderboard.yaml'))
-def test_get_raw_leaderboard(data_dir_mock, get_ids_mock, tmpdir):
+def test_get_raw_leaderboard(save_cache_mock, data_dir_mock, get_ids_mock, tmpdir):
```

Comment:
I'm not sure I see the reason to mock save_to_cache? Could you please explain this for me?

Reply:
I think get_raw_leaderboard will call get_leaderboard_row, which will call get_feedback. In some case get_feedback will cache updated results according to the new changes. So, I mocked those cases or it will return a FileNotFoundError when trying to save. 

Not sure this is the right way/good practice to write this test as I'm still learning.

Reply:
Thanks @khoapip. I guess my confusion arises from the fact that I don't really see why it actually needs to be mocked. We are using a tmpdir here, so all cache saves are made to this temporary directory as the data dir is redirected to the tmpdir. So I feel the test should pass regardless of whether save_to_cache is mocked. In parrticular, I'm struggling to see how a FileNotFoundError is being raised. Could you share a stack trace for this?

In general I was a bit suspicious about this code as having to mock a private function is usually a code smell. Private functions should have the right to change as they like, with no impact on the rest of the code (thats the ideal, at least). But if you ever have to mock it, this makes the test quite brittle to any such changes.

Reply:
Yeah, I was a bit hesitant to mock that private function too (and wasn't sure it was possible until I saw the test below that). I guess I didn't understand what "tmpdir" did, now that I read a bit more it seems it mock a temp directory to redirect all the files like you said. 

In the code, it was trying to open a file {tmpdir}/cache (in which the file cache doesn't exist). This is because we are calling get_raw_leaderboard without wraping it in cache(get_raw_leaderboard), so the "cache" folder is never created.  

We can add a  os.makedirs(cache_dir, exist_ok=True) check condition in both _load_from_cache and _save_to_cache, but not sure if you think this is a good practise. Let me know what you think!

![image](https://github.com/chai-research/chaiverse/assets/26734897/68b9bbe5-678a-4315-a84d-55f3df6e1454)

Reply:
I think the os.makedirs stuff makes sense, but looking more carefully at the code, I think the root cause is actually that guanaco_data_dir is mocked, which actually does do the os.makedirs stuff.

My gut says that the data_dir patch should be removed, and replaced with a monkey patch of the environment variable here: https://github.com/chai-research/chaiverse/blob/master/src/chaiverse/utils.py#L24C32-L24C42. (i.e just patch it to point to the tmpdir).

I believe you can do this with 

`with patch.dict(os.environ, {'GUANACO_DATA_DIR': tmpdir}, clear=True)`

You may even want to make that an auto fixture, which uses dependency injection to automatically pull the tmpdir, as so:

```python3
@pytest.fixture(autouse="session")
def mock_data_dir(tmpdir):
    with with patch.dict(os.environ, {'GUANACO_DATA_DIR': tmpdir}, clear=True):
        yield
```

Maybe something like that works?

Reply:
Hi @alexandred, thank you, it works! this seems to be a very elegant fix. Let me know what you think

Reply:
Looks great! Thanks again for your contribution :)